### PR TITLE
Improved error handling in deletion of delta snapshots during snapshot garbage collection

### LINKED ISSUE #652: [Enhancement] Error Handling in Delta Snapshot Deletion Process
**Enhancement (What you would like to be added):**

Improve the error handling mechanism in the delta snapshot deletion process [here](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/snapshot/snapshotter/garbagecollector.go#L226-L230). The proposed changes include:
1. Continuing with the deletion of the rest of the "deletable" delta snapshots even when an error is encountered, instead of halting the process.
2. Concatenating all errors encountered during the deletion process using `errors.Join()`, and returning this concatenated error at the end of the process.

>__Note__: This enhancement is based on the review comment in this https://github.com/gardener/etcd-backup-restore/pull/640#discussion_r1271316167.

**Motivation (Why is this needed?):**
The [current implementation](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/snapshot/snapshotter/garbagecollector.go#L226-L230) stops the garbage collection process of delta snapshots if there's an error with the deletion of a particular snapshot. This could be due to an intermittent error or an issue specific to a single snapshot. The proposed enhancement will ensure that a one-off error doesn't halt the entire process, while still effectively communicating to the function caller that there were errors in deleting delta snapshots.

**Approach/Hint to the implement solution (optional):**
1. Modify the deletion process to continue deleting the remaining snapshots even when an error is encountered.
2. Use the `errors.Join()` function to concatenate all errors encountered during the process.

### PR DESCRIPTION
**Pull request by Shreyas-s14**
**What this PR does / why we need it**:
This PR changes the error handling for deletion of delta snapshots by replacing the current implementation with concatenation of errors using the errors.Join() method upto a threshold of 5. The process of continues with the deletion of the rest of the "deletable" delta snapshots even when an error is encountered, instead of halting the process. When the number of errors becomes greater than the threshold, it errors out for the current cycle.

**Which issue(s) this PR fixes**:
Fixes #652 

**Release note**:

```other operator
Improved error handling in deletion of delta snapshots during snapshot garbage collection.
```

### CODE REVIEW

**Comment by gardener-robot-ci-1** [comment]
Thank you @Shreyas-s14 for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

File: pkg/snapshot/snapshotter/snapshotter_test.go [hunk 1]
```diff
@@ -528,6 +528,58 @@ var _ = Describe("Snapshotter", func() {
...
+
+						ssr, err := NewSnapshotter(logger, snapshotterConfig, store, etcdConnectionConfig, compressionConfig, healthConfig, snapstoreConfig)
+						Expect(err).ShouldNot(HaveOccurred())
+
+
+						// Threshold is set as 5
+						for i := len(list)-3 ; i > len(list)-9 ; i-- { // first 2 are deleted and next 5 return error while deleting
+							err = os.Remove(path.Join(list[i].Prefix, list[i].SnapName))
+							Expect(err).ShouldNot(HaveOccurred())
+						}
```
Comment (anveshreddy18) [inline pkg/snapshot/snapshotter/snapshotter_test.go:575]: Can the comment here be made better, when I first read it I felt the initial removal of objects itself errors out, but it should error out in the `ssr.GarbageCollectDeltaSnapshots()`. So moving this comment there would be better. I leave it to you, but it would be better to comment closer to the actual functionality.

File: pkg/snapshot/snapshotter/snapshotter_test.go [hunk 2]
```diff
@@ -528,6 +528,58 @@ var _ = Describe("Snapshotter", func() {
...
+						err = os.Remove(path.Join(list[3].Prefix, list[3].SnapName))
+						Expect(err).ShouldNot(HaveOccurred())
+						
+						err = os.Remove(path.Join(list[2].Prefix, list[2].SnapName))
+						Expect(err).ShouldNot(HaveOccurred())
+
+						deleted, err := ssr.GarbageCollectDeltaSnapshots(list)
+						Expect(deleted).ToNot(BeZero())
+						Expect(deleted).Should(Equal(6))
+						Expect(err).ToNot(BeNil())
```
Comment (anveshreddy18) [inline pkg/snapshot/snapshotter/snapshotter_test.go:557]:
```suggestion
						Expect(err).To(HaveOccurred())
```

This is to be consistent across the testcases, refererence: the other test case you've added.

**Review by anveshreddy18 — CHANGES_REQUESTED** [review]
Thanks for the PR @Shreyas-s14. Have few suggestions. PTAL

File: pkg/snapshot/snapshotter/snapshotter_test.go [hunk 3]
```diff
@@ -528,6 +528,51 @@ var _ = Describe("Snapshotter", func() {
 						Expect(len(list)).Should(Equal(3))
 					})
 				})
```
Comment (anveshreddy18) [inline pkg/snapshot/snapshotter/snapshotter_test.go:530]: You can also add a positive test case when there is no error in deleting snapshots.

File: pkg/snapshot/snapshotter/snapshotter_test.go [hunk 4]
```diff
@@ -528,6 +528,51 @@ var _ = Describe("Snapshotter", func() {
...
+							err := os.Remove(path.Join(list[i].Prefix, list[i].SnapName))
+							Expect(err).ShouldNot(HaveOccurred())
+						}
+
+						deleted, err := ssr.GarbageCollectDeltaSnapshots(list)
+						Expect(deleted).Should(Equal(6))
+						Expect(err).Should(HaveOccurred())
+					})
+				})
+				Context("When the number of errors while deleting are greater than the threshold", func() {
```
Comment (anveshreddy18) [inline pkg/snapshot/snapshotter/snapshotter_test.go:553]:
```suggestion
				Context("When the number of errors while deleting are greater than or equal to threshold", func() {
```

File: pkg/snapshot/snapshotter/garbagecollector.go [hunk 5]
```diff
@@ -85,14 +86,15 @@ func (ssr *Snapshotter) RunGarbageCollector(stopCh <-chan struct{}) {
 					eod          = now.Truncate(24 * time.Hour).Add(23 * time.Hour).Add(59 * time.Minute).Add(59 * time.Second)
 					trackingWeek = 0
 				)
+				// snapStream indicates the list of snapshot, where first snapshot is base/full snapshot followed by list of incremental snapshots based on it.
```
Comment (anveshreddy18) [inline pkg/snapshot/snapshotter/garbagecollector.go:89]: You can move this comment to above as `snapStream` is used there but defined here.

**Review by anveshreddy18 — CHANGES_REQUESTED** [review]

**Review by anveshreddy18 — APPROVED** [review]
Thanks for the PR @Shreyas-s14. Things are looking good. 
LGTM !!

File: pkg/snapshot/snapshotter/garbagecollector.go [hunk 6]
```diff
@@ -161,14 +163,14 @@ func (ssr *Snapshotter) RunGarbageCollector(stopCh <-chan struct{}) {
 			case brtypes.GarbageCollectionPolicyLimitBased:
 				// Delete delta snapshots in all snapStream but the latest one.
 				// Delete all snapshots beyond limit set by ssr.maxBackups.
-				for snapStreamIndex := 0; snapStreamIndex < len(snapStreamIndexList)-1; snapStreamIndex++ {
-					deletedSnap, err := ssr.GarbageCollectDeltaSnapshots(snapList[snapStreamIndexList[snapStreamIndex]:snapStreamIndexList[snapStreamIndex+1]])
+				for fullSnapshotIndex := 0; fullSnapshotIndex < len(fullSnapshotIndexList)-1; fullSnapshotIndex++ {
+					deletedSnap, err := ssr.GarbageCollectDeltaSnapshots(snapList[fullSnapshotIndexList[fullSnapshotIndex]:fullSnapshotIndexList[fullSnapshotIndex+1]])
```
Comment (renormalize) [inline pkg/snapshot/snapshotter/garbagecollector.go:167]: https://github.com/gardener/etcd-backup-restore/pull/793#discussion_r1818848135

File: pkg/snapshot/snapshotter/garbagecollector.go [hunk 7]
```diff
@@ -68,8 +70,8 @@ func (ssr *Snapshotter) RunGarbageCollector(stopCh <-chan struct{}) {
 				ssr.logger.Infof("GC: Total number garbage collected chunks: %d", chunksDeleted)
 			}
 
-			snapStreamIndexList := getSnapStreamIndexList(snapList)
-
+			fullSnapshotIndexList := getFullSnapshotIndexList(snapList)
+			// snapStream indicates the list of snapshot, where first snapshot is base/full snapshot followed by list of incremental snapshots based on it.
```
Comment (renormalize) [inline pkg/snapshot/snapshotter/garbagecollector.go:74]:
```suggestion
			// snapStream indicates a list of snapshots, where the first snapshot is base/full snapshot followed by a list of incremental snapshots based on it. Garbage collection is performed on one snapStream at a time.
```
Agree with @anveshreddy18's [review comment](https://github.com/gardener/etcd-backup-restore/pull/793/files#r1805250327), but I don't think the code comment is currently very useful in its state.

I don't particularly like that currently there is explanation about what `snapStream` is, but there is no `snapStream` token in the function where the explanation is present.
There is no variable/type in the function with `snapStream` in it. Unless someone searches for that word in the rest of the file, the comment causes more confusion.

See the suggestion in https://github.com/gardener/etcd-backup-restore/pull/793#discussion_r1818848135, would make the above suggested code read better.

**Review by renormalize — CHANGES_REQUESTED** [review]
Thanks for the PR @Shreyas-s14!
I'd like to suggest a few changes.

File: pkg/snapshot/snapshotter/snapshotter_test.go [hunk 8]
```diff
@@ -576,8 +576,7 @@ var _ = Describe("Snapshotter", func() {
 						Expect(err).ShouldNot(HaveOccurred())
 
 
-						// Threshold is set as 5
-						for i := len(list)-3 ; i > len(list)-9 ; i-- { // 5 snapshots from the list of snapshots to be deleted, are removed before passing it to the function ssr.GarbageCollectDeltaSnapshots. This will cause an error when the function tries to delete them.
+						for i := len(list)-3 ; i > len(list)- 3 - ErrorThreshold-1 ; i-- { // 5 snapshots from the list of snapshots to be deleted, are removed before passing it to the function ssr.GarbageCollectDeltaSnapshots. This will cause an error when the function tries to delete them.
```
Comment (renormalize) [inline pkg/snapshot/snapshotter/snapshotter_test.go:579]: - The formatting looks odd in the loop condition. 
Do you have `gofmt` configured? Please run `gofmt`.
- Can you move the comment above the for loop? It's quite hard to follow with so many characters in one line.
- Update the comment to not mention `5 snapshots from the list of snapshots`. Mention the threshold.
- Mention that the function will error when the number of individual errors seen meets the threshold set.

**Review by renormalize — CHANGES_REQUESTED** [review]

File: pkg/snapshot/snapshotter/garbagecollector.go [hunk 9]
```diff
@@ -89,13 +90,16 @@ func (ssr *Snapshotter) RunGarbageCollector(stopCh <-chan struct{}) {
...
 				// Here we start processing from second last snapstream, because we want to keep last snapstream
 				// including delta snapshots in it.
-				for fullSnapshotIndex := len(fullSnapshotIndexList) - 1; fullSnapshotIndex > 0; fullSnapshotIndex-- {
-					snap := snapList[fullSnapshotIndexList[fullSnapshotIndex]]
-					nextSnap := snapList[fullSnapshotIndexList[fullSnapshotIndex-1]]
+				for fullSnapshotIndex := 0; fullSnapshotIndex < len(fullSnapshotIndexList) - 1; fullSnapshotIndex++ {
+					descIndex := len(fullSnapshotIndexList) - 1 - fullSnapshotIndex
+    				nextDescIndex := descIndex - 1
+					snap := snapList[fullSnapshotIndexList[descIndex]]
+					nextSnap := snapList[fullSnapshotIndexList[nextDescIndex]]
```
[the remark this answers is not in the document]
Comment (renormalize) [reply pkg/snapshot/snapshotter/garbagecollector.go]: The tests are not written in a way which is agnostic to the order in which the loop is run, oddly. When the order is changed of the loop, the tests fail.

We've decided to correct this later.

File: pkg/snapshot/snapshotter/garbagecollector.go [hunk 10]
```diff
@@ -5,6 +5,8 @@
 package snapshotter
 
 import (
+	"errors"
+	// "fmt"
```
Comment (renormalize) [inline pkg/snapshot/snapshotter/garbagecollector.go:9]: This commented import can be removed.

File: pkg/snapshot/snapshotter/garbagecollector.go [hunk 11]
```diff
@@ -16,6 +18,7 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 )
 
+const DeltaSnapshotGCErrorThreshold = 5
```
Comment (renormalize) [inline pkg/snapshot/snapshotter/garbagecollector.go:21]: A comment has to be added to the exported `const` for `check` to pass in CI. Please add a comment.

**Review by renormalize — CHANGES_REQUESTED** [review]

**Review by renormalize — APPROVED** [review]
